### ISSUE TITLE
Create a function producing a summary of all values for a table

### ISSUE THREAD

**Issue by pedrogit** [edwardsmarc/CASFRI#235]

**Comment by edwardsmarc** [comment]
This is a good idea and I believe @prvernier has been doing something similar with R to summarize source tables.

Once the summary table is made, would you then go and create histograms as a second step for the integer/float attributes?

If the purpose is to support exploration of source data when developing translation rules and getting to know a new inventory, we might want to consider if sql is the best language for this? @prvernier for example will be much faster doing this in R.

**Comment by pedrogit** [comment]
> Once the summary table is made, would you then go and create histograms as a second step for the integer/float attributes?

You mean graphical histograms? No...

> If the purpose is to support exploration of source data when developing translation rules and getting to know a new inventory, we might want to consider if sql is the best language for this? @prvernier for example will be much faster doing this in R.

The purpose is to be able to explore both target tables and source tables. Did you do some tests to conclude that R would be faster? I don't think developing R tools is in accordance with the project specifications...

**Comment by edwardsmarc** [comment]
> You mean graphical histograms? No...

Yes I meant histogram graphs. I understand now that you are meaning histograms to be just the binned groups shown in your example output table.

> Did you do performance tests to conclude that R would be faster?

I meant it is faster for @prvernier to explore the data using R than learning a new language for data exploration (sql). I'm not suggesting R is faster to create the summaries, I don't think processing speed is an issue here.

> I don't think developing R tools is in accordance with the project specifications...

None of the project specifications mention developing sql tools for data exploration. In my mind the user (in this case @prvernier) can explore the data in whatever way they like. The deliverable is the translation rules as detailed in the translation tables. The method used to generate those rules is not a deliverable. However, regardless of the method used, now we have @prvernier and @MelinaHoule both working on rules it would make sense for us to have a standardized method for exploring data (maybe that is indeed sql, I don't know).

**Comment by pedrogit** [comment]
> None of the project specifications mention developing sql tools for data exploration. In my mind the user (in this case @prvernier) can explore the data in whatever way they like. The deliverable is the translation rules as detailed in the translation tables. The method used to generate those rules is not a deliverable. However, regardless of the method used, now we have @prvernier and @MelinaHoule both working on rules it would make sense for us to have a standardized method for exploring data (maybe that is indeed sql, I don't know).

Right. But I prefer to stick to SQL...

**Comment by prvernier** [comment]
For exploring and summarizing tables, R already has a lot of simple and useful tools, some of which I use e.g., dfSummary. And reading and querying postgreSQL tables is also simple and fast. I suppose, if I was using SQL for everything, I would be interested in additional functions that would facilitate my work. Maybe third party users would too.

**Comment by pedrogit** [comment]
Could you provide a R code snippet that does a summary of casfri50.cas_all from the database?

**Comment by prvernier** [comment]
library(rpostgis)
library(summarytools)

con = dbConnect(RPostgreSQL::PostgreSQL(), dbname="casfri50_pierrev", host="localhost", port=5432, user="postgres", password="1postgres")
bc10 = dbGetQuery(con, "SELECT * FROM rawfri.bc10 LIMIT 10000")
dbDisconnect(con)

dfSummary(bc10)

**Comment by prvernier** [comment]
Note that was just for BC10, but the same could be do for casfri50.cas_all.

**Comment by pedrogit** [comment]
Neat! A quick run on casfr50_test found that:

- -9999 was assigned to stand_structure when it's a text attribute
- there are 57 TRANSLATION_ERROR for stand_structure
- there are 57 TRANSLATION_ERROR for num_of_layers. (Currently being revisited by Mark)
- many NOT_IN_SET. Many of them should probably be added to the validation rules
- dist_year_1,2,3 shows 0s and other very stange values
- some strange 0% species_pct and crown_closure_upper
- should not structure_per be a true percentage?
- some overprecise height_upper and height_lower

**Comment by edwardsmarc** [comment]
If we don't have an sql function for this we can do it in R to validate the final casfri translations

**Comment by edwardsmarc** [comment]
@pedrogit can we use tt_summarize() to create summaries from the final CAS db for each inventory?
This could be useful for checking the translated values make sense. For example catching cases where species percent were not multiplied by 10.

**Comment by pedrogit** [comment]
tt_summarize() was never written because there was an easy R snippet of code to do the same thing. tt_summarize is also a hard function to write.

**Comment by edwardsmarc** [comment]
Function and output has been made.

Output needs to be reviewed.
One thing to check for is cases where we should catch zero values as nulls using notMatchList(). These would show up as NOT_IN_SET or OUT_OF_RANGE in CASFRI.

**Comment by pedrogit** [comment]
At some point we will have to convert the summary_statistic R code to SQL and have the mechanism to compare the results with archived ones.
